### PR TITLE
Change annotations to use a NamedTuple

### LINKED ISSUE #55245: AnnotatedString constructor too restrictive
```julia
julia> anns = [:x => "a", :y => 1]

# this works:
julia> Base.AnnotatedString("abc def", [(i:i, anns[i]) for i in 1:2])
"abc def"

# but after replacing 2 with 0 doesn't:
julia> Base.AnnotatedString("abc def", [(i:i, anns[i]) for i in 1:0])
ERROR: MethodError: no method matching Base.AnnotatedString(::String, ::Vector{Tuple{UnitRange{Int64}, Any}})
```
I think the constructor type constraints are too restrictive.

### LINKED ISSUE #55249: Add meaningful names to AnnotatedString annotations to make their usage understandable
This is IMO quite unintuitive and confusing to read:
```julia
julia> s = Base.AnnotatedString("this is an example annotated string", [(1:18, :A => 111), (12:28, :B => 222), (18:35, :C => 333)])

julia> Base.annotations(s)[2][1]
12:28

julia> Base.annotations(s)[2][2][1]  # or Base.annotations(s)[2][2].first
:B

julia> Base.annotations(s)[2][2][2]  # or Base.annotations(s)[2][2].second
222
```
Would be much nicer and self-explanatory to use meaningful names instead of numbers:
```julia
julia> Base.annotations(s)[2].region
12:28

julia> Base.annotations(s)[2].label
:B

julia> Base.annotations(s)[2].value
222
```

### PR DESCRIPTION
**Pull request by tecosaur**
Due to popular demand, the type of annotations is to be changed from a `Tuple{UnitRange{Int}, Pair{Symbol, Any}}` to a `NamedTuple{(:region, :label, :value), Tuple{UnitRange{Int}, Symbol,
Any}}`.

This requires the expected code churn to `strings/annotated.jl`, and some changes to the StyledStrings and JuliaSyntaxHighlighting libraries.

Closes #55249 and closes #55245.

### CODE REVIEW

**Comment by tecosaur** [comment]
Looks like all the tests are happy, it's just a matter of updating the docs(trings).

**Comment by tecosaur** [comment]
Markdown also needed to be updated.

**Comment by tecosaur** [comment]
...and some more tests that need updating

**Comment by tecosaur** [comment]
Hopefully that does it :crossed_fingers:

**Comment by KristofferC** [comment]
CI seems to fail with some `===` checks.

**Comment by tecosaur** [comment]
It occurs to me that we'll want to backport the StyledStrings version bump, but not the JuliaSyntaxHighlighing bump. With this in mind, I'll split the bump commit when I figure out the method ambiguity (or somebody else proposes a resolution).

**Comment by tecosaur** [comment]
I think I've sorted out the method ambiguity that CI was complaining about, and the StyledStrings/JuliaSyntaxHighlighting bumps have been split up to make backporting easier.

**Comment by tecosaur** [comment]
CI looks promising! The only error I see is a generic windows bootstrap error.

**Comment by tecosaur** [comment]
With tests passing, do we want to go ahead and merge this, or is there anything else I should do here?

**Comment by tecosaur** [comment]
I think this should be good to merge, but I'd appreciate it if somebody else could take a look and do so if this looks good to them.

**Comment by tecosaur** [comment]
I've just rebased + bumped the bump to incorporate https://github.com/JuliaLang/StyledStrings.jl/pull/81.

**Comment by tecosaur** [comment]
Test failures seem unrelated (from Pkg)

**Comment by tecosaur** [comment]
Ta, there we go.

I think it would be best if this could be merged in time to make it into 1.11, but I don't think Kristoffer has the bandwidth to review this and I'm generally uncomfortable with being the only person involved in approving my own PRs.

**Comment by tecosaur** [comment]
Once again, test failures seem unrelated (from Pkg, but different this time).

**Comment by tecosaur** [comment]
@aplavin @fatteneder since both of you were involved in the linked issues, might either of you be interested in reviewing this?

File: base/strings/annotated.jl
Comment (tecosaur) [reply base/strings/annotated.jl:3]: Thanks for giving this a look. This is mainly to avoid writing `NamedTuple{(:region, :label, :value), Tuple{UnitRange{Int}, Symbol, Any}}` again and again in a bunch of places, but for checking if something `isa` region annotation we would need to use `<:Tuple{., ., <:Any}`.

Maybe this should also be defined as a `const` for convenience? Maybe after doing so the values should be swapped?

I'm not sure, but at least I can say that this is very much a private internal (just with a general name), and so we're not locking ourselves into anything here.

Glad to hear nothing else looks notable, I didn't anticipate any major comments but I do like other people reviewing work I do in `JuliaLang/julia` before it's merged.

**Comment by tecosaur** [comment]
I'd love it if another committer could have a look at and merge this, but given how impending 1.11 is and that I think this should make it in, I will just merge this myself shortly if there are no objections.

File: base/strings/annotated.jl
Comment (fatteneder) [inline base/strings/annotated.jl:375]:
```suggestion
        Vector{$RegionAnnotation}
```

File: base/strings/annotated.jl
Comment (fatteneder) [inline base/strings/annotated.jl:539]:
```suggestion
    _clear_annotations_in_region!(annotations::Vector{$RegionAnnotation}, span::UnitRange{Int})
```

File: base/strings/annotated.jl
Comment (fatteneder) [inline base/strings/annotated.jl:579]:
```suggestion
    _insert_annotations!(io::AnnotatedIOBuffer, annotations::Vector{$RegionAnnotation}, offset::Int = position(io))
```

**Review by fatteneder — APPROVED** [review]
just cosmetics for some doc strings

**Comment by tecosaur** [comment]
Thanks Florian, I've just applied those. I'll (hopefully) be asleep when CI finished running, but if you'd be happy to merge this based on your look through the code that would be great.

If not, I'll merge this within ~24h to try to squeeze it into 1.11.

**Comment by tecosaur** [comment]
Ah well, 1.11.0 has been cut now. No rush necessitating me to merge it myself any more.

**Review by topolarity — APPROVED** [review]
Changes are looking good here - just a couple of small infer-ability issues that jumped out to me.

File: base/strings/annotated.jl
Comment (topolarity) [reply base/strings/annotated.jl:338]: Took a stab at improving this in general in: https://github.com/JuliaLang/julia/pull/56034

Comment (tecosaur) [reply base/strings/annotated.jl:338]: Would you say this is still worth doing?

Comment (tecosaur) [reply base/strings/annotated.jl:338]: I suppose I might as well :thinking:

**Comment by topolarity** [comment]
@tecosaur OK if I merge here?

**Comment by tecosaur** [comment]
That would be great by me!

**Pull request merged**
